Quarkus won't start if both the micrometer extension and the vertx-web extension are enabled

**Describe the bug**

```
Caused by: java.lang.ExceptionInInitializerError
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:416)
        at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:196)
        at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:234)
        at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:512)
        at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:545)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$8(ClassBasedTestDescriptor.java:368)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:368)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:192)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:78)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:136)
        ... 34 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:200)
        ... 46 more
Caused by: java.lang.NoClassDefFoundError: io/quarkus/resteasy/common/runtime/ResteasyContextProvider
        at io.quarkus.deployment.steps.SmallRyeContextPropagationProcessor$buildStatic-850855720.deploy_0(SmallRyeContextPropagationProcessor$buildStatic-850855720.zig:79)
        at io.quarkus.deployment.steps.SmallRyeContextPropagationProcessor$buildStatic-850855720.deploy(SmallRyeContextPropagationProcessor$buildStatic-850855720.zig:40)
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:136)
        ... 46 more
Caused by: java.lang.ClassNotFoundException: io.quarkus.resteasy.common.runtime.ResteasyContextProvider
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:602)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:406)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:363)
```

**To Reproduce**

Add the micrometer extension to the vertx-web integration-test.

Down the rabbit hole a bit farther: 

```
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:200)
        ... 46 more
Caused by: java.lang.NoClassDefFoundError: io/quarkus/resteasy/common/runtime/ResteasyInjectorFactoryRecorder
        at io.quarkus.deployment.steps.ResteasyCommonProcessor$setupResteasyInjection-1799175235.deploy_0(ResteasyCommonProcessor$setupResteasyInjection-1799175235.zig:97)
        at io.quarkus.deployment.steps.ResteasyCommonProcessor$setupResteasyInjection-1799175235.deploy(ResteasyCommonProcessor$setupResteasyInjection-1799175235.zig:40)
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:169)
        ... 46 more
Caused by: java.lang.ClassNotFoundException: io.quarkus.resteasy.common.runtime.ResteasyInjectorFactoryRecorder
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:602)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:406)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:363)
        ... 49 more
```

This is ultimately due to the use of io.quarkus.resteasy.common.spi.ResteasyJaxrsProviderBuildItem to register the container filter. To use that build item, I need this: 

```
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-resteasy-server-common-deployment</artifactId>
        </dependency>
```

And that makes services from resteasy visible to service factories, which then attempt to instantiate the things.. and it doesn't end.

--- a/extensions/micrometer/deployment/pom.xml
+++ b/extensions/micrometer/deployment/pom.xml
@@ -40,7 +40,7 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-resteasy-server-common-deployment</artifactId>
+            <artifactId>quarkus-resteasy-common-spi</artifactId>
         </dependency>
 
         <!-- Registry providers (optional) -->

--- a/extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/binder/VertxBinderProcessor.java
+++ b/extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/binder/VertxBinderProcessor.java
@@ -5,6 +5,9 @@ import java.util.function.BooleanSupplier;
 import javax.interceptor.Interceptor;
 
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
+import io.quarkus.deployment.Capabilities;
+import io.quarkus.deployment.Capability;
+import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.annotations.ExecutionTime;
 import io.quarkus.deployment.annotations.Record;
@@ -31,18 +34,26 @@ public class VertxBinderProcessor {
         }
     }
 
+    @BuildStep(onlyIf = { VertxBinderEnabled.class })
+    ResteasyJaxrsProviderBuildItem enableResteasySupport(Capabilities capabilities,
+            BuildProducer<AdditionalBeanBuildItem> additionalBeans) {
+        if (!capabilities.isPresent(Capability.RESTEASY)) {
+            return null;
+        }
+
+        additionalBeans.produce(AdditionalBeanBuildItem.builder()
+                .addBeanClass(VertxMeterBinderContainerFilter.class)
+                .setUnremovable().build());
+
+        return new ResteasyJaxrsProviderBuildItem(VertxMeterBinderContainerFilter.class.getName());
+    }
+
     @BuildStep(onlyIf = VertxBinderEnabled.class)
     AdditionalBeanBuildItem createVertxAdapters() {
         // Add Vertx meter adapters
         return AdditionalBeanBuildItem.builder()
                 .addBeanClass(VertxMeterBinderAdapter.class)
-                .addBeanClass(VertxMeterBinderContainerFilter.class)
                 .setUnremovable().build();
-    }
-
-    @BuildStep(onlyIf = VertxBinderEnabled.class)
-    ResteasyJaxrsProviderBuildItem createVertxFilters() {
-        return new ResteasyJaxrsProviderBuildItem(VertxMeterBinderContainerFilter.class.getName());
     }
 
     @BuildStep(onlyIf = VertxBinderEnabled.class)

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/vertx/RequestMetric.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/vertx/RequestMetric.java
@@ -1,17 +1,34 @@
 package io.quarkus.micrometer.runtime.binder.vertx;
 
-import java.util.HashMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Pattern;
 
+import io.micrometer.core.instrument.Timer;
 import io.vertx.core.Context;
 import io.vertx.ext.web.RoutingContext;
 
-public class RequestMetric extends HashMap<String, Object> {
+public class RequestMetric {
     static final String METRICS_CONTEXT = "HTTP_REQUEST_METRICS_CONTEXT";
-
     static final String HTTP_REQUEST_PATH = "HTTP_REQUEST_PATH";
-    static final String HTTP_REQUEST_SAMPLE = "HTTP_REQUEST_SAMPLE";
+    static final String HTTP_REQUEST_PATH_MATCHED = "HTTP_REQUEST_MATCHED_PATH";
+    static final Pattern VERTX_ROUTE_PARAM = Pattern.compile("^:(.*)$");
+
+    /** Cache of vert.x resolved paths: /item/:id --> /item/{id} */
+    final static ConcurrentHashMap<String, String> vertxRoutePath = new ConcurrentHashMap<>();
 
     volatile RoutingContext routingContext;
+
+    /** Do not measure requests until/unless a uri path is set */
+    boolean measure = false;
+
+    /** URI path used as a tag value for non-error requests */
+    String path;
+
+    /** True IFF the path was revised by a matcher expression */
+    boolean pathMatched = false;
+
+    /** Store the sample used to measure the request */
+    Timer.Sample sample;
 
     /**
      * Stash the RequestMetric in the Vertx Context
@@ -42,15 +59,33 @@ public class RequestMetric extends HashMap<String, Object> {
         return null;
     }
 
-    <T> T getFromRoutingContext(String key) {
-        if (routingContext != null) {
-            return routingContext.get(key);
+    String getHttpRequestPath() {
+        // Vertx binder configuration, see VertxMetricsTags
+        if (pathMatched) {
+            return path;
         }
-        return null;
-    }
-
-    <T> T getValue(String key) {
-        Object o = this.get(key);
-        return o == null ? null : (T) o;
+        if (routingContext != null) {
+            // JAX-RS or Servlet container filter
+            String rcPath = routingContext.get(HTTP_REQUEST_PATH);
+            if (rcPath != null) {
+                return rcPath;
+            }
+            // vertx-web or reactive route
+            String matchedPath = routingContext.currentRoute().getPath();
+            if (matchedPath != null) {
+                if (matchedPath.contains(":")) {
+                    // Convert /item/:id to /item/{id} and save it for next time
+                    matchedPath = vertxRoutePath.computeIfAbsent(matchedPath, k -> {
+                        String segments[] = k.split("/");
+                        for (int i = 0; i < segments.length; i++) {
+                            segments[i] = VERTX_ROUTE_PARAM.matcher(segments[i]).replaceAll("{$1}");
+                        }
+                        return String.join("/", segments);
+                    });
+                }
+                return matchedPath;
+            }
+        }
+        return path;
     }
 }

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/vertx/VertxHttpServerMetrics.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/vertx/VertxHttpServerMetrics.java
@@ -2,9 +2,11 @@ package io.quarkus.micrometer.runtime.binder.vertx;
 
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
 
 import org.jboss.logging.Logger;
 
@@ -34,7 +36,7 @@ public class VertxHttpServerMetrics extends VertxTcpMetrics
     static final Logger log = Logger.getLogger(VertxHttpServerMetrics.class);
 
     final List<Pattern> ignorePatterns;
-    final List<Pattern> matchPatterns;
+    final Map<Pattern, String> matchPatterns;
 
     final String nameWebsocketConnections;
     final String nameHttpServerPush;
@@ -58,12 +60,23 @@ public class VertxHttpServerMetrics extends VertxTcpMetrics
 
         if (config.matchPatterns.isPresent()) {
             List<String> stringPatterns = config.matchPatterns.get();
-            matchPatterns = new ArrayList<>(stringPatterns.size());
+            matchPatterns = new HashMap<>(stringPatterns.size());
             for (String s : stringPatterns) {
-                matchPatterns.add(Pattern.compile(s));
+                int pos = s.indexOf("=");
+                if (pos > 0 && s.length() > 2) {
+                    String pattern = s.substring(0, pos);
+                    String replacement = s.substring(pos + 1);
+                    try {
+                        matchPatterns.put(Pattern.compile(pattern), replacement);
+                    } catch (PatternSyntaxException pse) {
+                        log.errorf("Invalid pattern in replacement string (%s=%s): %s", pattern, replacement, pse);
+                    }
+                } else {
+                    log.errorf("Invalid pattern in replacement string (%s). Should be pattern=replacement", s);
+                }
             }
         } else {
-            matchPatterns = Collections.emptyList();
+            matchPatterns = Collections.emptyMap();
         }
     }
 
@@ -80,10 +93,10 @@ public class VertxHttpServerMetrics extends VertxTcpMetrics
     public RequestMetric responsePushed(Map<String, Object> socketMetric, HttpMethod method, String uri,
             HttpServerResponse response) {
         RequestMetric requestMetric = new RequestMetric();
-        String path = VertxMetricsTags.parseUriPath(matchPatterns, ignorePatterns, uri);
-        if (path != null) {
+        VertxMetricsTags.parseUriPath(requestMetric, matchPatterns, ignorePatterns, uri);
+        if (requestMetric.measure) {
             registry.counter(nameHttpServerPush, Tags.of(
-                    VertxMetricsTags.uri(path, response.getStatusCode()),
+                    VertxMetricsTags.uri(requestMetric.path, response.getStatusCode()),
                     VertxMetricsTags.method(method),
                     VertxMetricsTags.outcome(response),
                     VertxMetricsTags.status(response.getStatusCode())))
@@ -107,17 +120,16 @@ public class VertxHttpServerMetrics extends VertxTcpMetrics
         RequestMetric requestMetric = new RequestMetric();
         RequestMetric.setRequestMetric(Vertx.currentContext(), requestMetric);
 
-        String path = VertxMetricsTags.parseUriPath(matchPatterns, ignorePatterns, request.uri());
-        if (path != null) {
-            // Pre-add the request method tag to the sample
-            requestMetric.put(RequestMetric.HTTP_REQUEST_SAMPLE,
-                    Timer.start(registry).tags(Tags.of(VertxMetricsTags.method(request.method()))));
+        // evaluate and remember the path to monitor for use later (maybe a 404 or redirect..)
+        VertxMetricsTags.parseUriPath(requestMetric, matchPatterns, ignorePatterns, request.path());
+        if (requestMetric.measure) {
+            // If we're measuring this request, create/remember the sample
+            requestMetric.sample = Timer.start(registry).tags(Tags.of(
+                    VertxMetricsTags.method(request.method())));
 
-            // remember the path to monitor for use later (maybe a 404 or redirect..)
-            requestMetric.put(RequestMetric.HTTP_REQUEST_PATH, path);
+            log.debugf("requestBegin %s: %s, %s", requestMetric.path, socketMetric, requestMetric);
         }
 
-        log.debugf("requestBegin %s: %s, %s, %s", request.uri(), socketMetric, requestMetric, request.headers());
         return requestMetric;
     }
 
@@ -200,19 +212,13 @@ public class VertxHttpServerMetrics extends VertxTcpMetrics
         if (metricsContext == null) {
             return null;
         }
-        return metricsContext.getValue(RequestMetric.HTTP_REQUEST_SAMPLE);
+        return metricsContext.sample;
     }
 
     private String getServerRequestPath(RequestMetric metricsContext) {
         if (metricsContext == null) {
             return null;
         }
-
-        String path = metricsContext.getFromRoutingContext(RequestMetric.HTTP_REQUEST_PATH);
-        if (path != null) {
-            log.debugf("Using path from routing context %s", path);
-            return path;
-        }
-        return metricsContext.getValue(RequestMetric.HTTP_REQUEST_PATH);
+        return metricsContext.getHttpRequestPath();
     }
 }

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/vertx/VertxMeterBinderContainerFilter.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/vertx/VertxMeterBinderContainerFilter.java
@@ -20,20 +20,26 @@ public class VertxMeterBinderContainerFilter implements ContainerRequestFilter {
     @Override
     public void filter(final ContainerRequestContext requestContext) {
         RoutingContext routingContext = CDI.current().select(CurrentVertxRequest.class).get().getCurrent();
+
+        // bail early if we have no routing context, or if path munging isn't necessary
+        if (routingContext == null || routingContext.get(RequestMetric.HTTP_REQUEST_PATH_MATCHED) != null) {
+            return;
+        }
+
         UriInfo info = requestContext.getUriInfo();
+        String path = info.getPath();
 
         MultivaluedMap<String, String> pathParameters = info.getPathParameters();
-        if (!pathParameters.isEmpty() && routingContext != null) {
+        if (!pathParameters.isEmpty()) {
             // Replace parameter values in the URI with {key}: /items/123 -> /items/{id}
-            String path = info.getPath();
             for (Map.Entry<String, List<String>> entry : pathParameters.entrySet()) {
                 for (String value : entry.getValue()) {
                     path = path.replace(value, "{" + entry.getKey() + "}");
                 }
             }
-
             log.debugf("Saving parameterized path %s in %s", path, routingContext);
-            routingContext.put(RequestMetric.HTTP_REQUEST_PATH, path);
         }
+
+        routingContext.put(RequestMetric.HTTP_REQUEST_PATH, path);
     }
 }

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/vertx/VertxMeterFilter.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/vertx/VertxMeterFilter.java
@@ -1,7 +1,5 @@
 package io.quarkus.micrometer.runtime.binder.vertx;
 
-import org.jboss.logging.Logger;
-
 import io.vertx.core.Context;
 import io.vertx.core.Handler;
 import io.vertx.core.Vertx;
@@ -14,17 +12,20 @@ import io.vertx.ext.web.RoutingContext;
  * Reference: https://github.com/eclipse-vertx/vert.x/issues/3579
  */
 public class VertxMeterFilter implements Handler<RoutingContext> {
-    private static final Logger log = Logger.getLogger(VertxMeterFilter.class);
-
     @Override
-    public void handle(RoutingContext event) {
+    public void handle(RoutingContext routingContext) {
         final Context context = Vertx.currentContext();
-        log.debugf("Handling event %s with context %s", event, context);
-
         RequestMetric requestMetric = RequestMetric.retrieveRequestMetric(context);
+
         if (requestMetric != null) {
-            requestMetric.routingContext = event;
+            requestMetric.routingContext = routingContext;
+
+            // remember if we can skip path munging --> @see VertxMeterBinderContainerFilter
+            if (requestMetric.pathMatched) {
+                routingContext.put(RequestMetric.HTTP_REQUEST_PATH_MATCHED, true);
+            }
         }
-        event.next();
+        routingContext.next();
     }
+
 }

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/vertx/VertxMetricsTags.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/vertx/VertxMetricsTags.java
@@ -1,6 +1,7 @@
 package io.quarkus.micrometer.runtime.binder.vertx;
 
 import java.util.List;
+import java.util.Map;
 import java.util.regex.Pattern;
 
 import org.jboss.logging.Logger;
@@ -110,9 +111,10 @@ public class VertxMetricsTags {
      * Extract the path out of the uri. Return null if the path should be
      * ignored.
      */
-    static String parseUriPath(List<Pattern> matchPattern, List<Pattern> ignorePatterns, String uri) {
+    static void parseUriPath(RequestMetric requestMetric, Map<Pattern, String> matchPattern, List<Pattern> ignorePatterns,
+            String uri) {
         if (uri == null) {
-            return null;
+            return;
         }
 
         String path = "/" + extractPath(uri);
@@ -122,15 +124,20 @@ public class VertxMetricsTags {
         if (path.isEmpty()) {
             path = "/";
         }
+        requestMetric.path = path;
+        for (Map.Entry<Pattern, String> mp : matchPattern.entrySet()) {
+            requestMetric.path = mp.getKey().matcher(requestMetric.path).replaceAll(mp.getValue());
+        }
+        requestMetric.pathMatched = !path.equals(requestMetric.path);
 
         // Compare path against "ignore this path" patterns
         for (Pattern p : ignorePatterns) {
             if (p.matcher(path).matches()) {
                 log.debugf("Path %s ignored; matches pattern %s", uri, p.pattern());
-                return null;
+                return;
             }
         }
-        return path;
+        requestMetric.measure = true;
     }
 
     private static String extractPath(String uri) {

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/config/runtime/VertxConfig.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/config/runtime/VertxConfig.java
@@ -10,18 +10,28 @@ import io.quarkus.runtime.annotations.ConfigRoot;
 @ConfigRoot(name = "micrometer.binder.vertx", phase = ConfigPhase.RUN_TIME)
 public class VertxConfig {
     /**
-     * Comma-separated case-sensitive list of regular expressions defining Paths
-     * that should be matched and used as tags. By default, the first path
-     * segment will be used. This default behavior will also apply to any
-     * URI path that is found (2xx or 5xx) but doesn't match elements
-     * in this list.
+     * Comma-separated list of regular expressions used to specify uri
+     * labels in http metrics.
+     *
+     * Vertx instrumentation will attempt to transform parameterized
+     * resource paths, `/item/123`, into a generic form, `/item/{id}`,
+     * to reduce the cardinality of uri label values.
+     *
+     * Patterns specified here will take precedence over those computed
+     * values.
+     *
+     * For example, if `/item/\\d+=/item/custom` is specified in this list,
+     * a request to a matching path (`/item/123`) will use the specified
+     * replacement value (`/item/custom`) as the value for the uri label.
+     *
+     * @asciidoclet
      */
     @ConfigItem
     public Optional<List<String>> matchPatterns = Optional.empty();
 
     /**
-     * Comma-separated case-sensitive list of regular expressions defining Paths
-     * that should be ignored / not measured.
+     * Comma-separated list of regular expressions defining uri paths
+     * that should be ignored (not measured).
      */
     @ConfigItem
     public Optional<List<String>> ignorePatterns = Optional.empty();

--- a/extensions/micrometer/runtime/src/test/java/io/quarkus/micrometer/runtime/binder/vertx/RequestMetricTest.java
+++ b/extensions/micrometer/runtime/src/test/java/io/quarkus/micrometer/runtime/binder/vertx/RequestMetricTest.java
@@ -1,0 +1,54 @@
+package io.quarkus.micrometer.runtime.binder.vertx;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.DisabledOnJre;
+import org.junit.jupiter.api.condition.JRE;
+import org.mockito.Mockito;
+
+import io.vertx.ext.web.Route;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Disabled on Java 8 because of Mocks
+ */
+@DisabledOnJre(JRE.JAVA_8)
+public class RequestMetricTest {
+
+    @Test
+    public void testReturnPathFromHttpRequestPath() {
+        RequestMetric requestMetric = new RequestMetric();
+        requestMetric.routingContext = Mockito.mock(RoutingContext.class);
+
+        Mockito.when(requestMetric.routingContext.get(RequestMetric.HTTP_REQUEST_PATH))
+                .thenReturn("/item/{id}");
+
+        Assertions.assertEquals("/item/{id}", requestMetric.getHttpRequestPath());
+    }
+
+    @Test
+    public void testReturnPathFromRoutingContext() {
+        RequestMetric requestMetric = new RequestMetric();
+        requestMetric.routingContext = Mockito.mock(RoutingContext.class);
+        Route currentRoute = Mockito.mock(Route.class);
+
+        Mockito.when(requestMetric.routingContext.currentRoute()).thenReturn(currentRoute);
+        Mockito.when(currentRoute.getPath()).thenReturn("/item");
+
+        Assertions.assertEquals("/item", requestMetric.getHttpRequestPath());
+    }
+
+    @Test
+    public void testReturnGenericPathFromRoutingContext() {
+        RequestMetric requestMetric = new RequestMetric();
+        requestMetric.routingContext = Mockito.mock(RoutingContext.class);
+        Route currentRoute = Mockito.mock(Route.class);
+
+        Mockito.when(requestMetric.routingContext.currentRoute()).thenReturn(currentRoute);
+        Mockito.when(currentRoute.getPath()).thenReturn("/item/:id");
+
+        Assertions.assertEquals("/item/{id}", requestMetric.getHttpRequestPath());
+        // Make sure conversion is cached
+        Assertions.assertEquals("/item/{id}", RequestMetric.vertxRoutePath.get("/item/:id"));
+    }
+}

--- a/extensions/micrometer/runtime/src/test/java/io/quarkus/micrometer/runtime/binder/vertx/VertxHttpServerMetricsTest.java
+++ b/extensions/micrometer/runtime/src/test/java/io/quarkus/micrometer/runtime/binder/vertx/VertxHttpServerMetricsTest.java
@@ -1,0 +1,41 @@
+package io.quarkus.micrometer.runtime.binder.vertx;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+import io.quarkus.micrometer.runtime.config.runtime.VertxConfig;
+
+public class VertxHttpServerMetricsTest {
+
+    @Test
+    public void testHttpServerMetricsIgnorePatterns() {
+        VertxConfig runtimeConfig = new VertxConfig();
+        runtimeConfig.ignorePatterns = Optional.of(new ArrayList<>(Arrays.asList("/item/.*")));
+        VertxHttpServerMetrics metrics = new VertxHttpServerMetrics(new SimpleMeterRegistry(), runtimeConfig);
+
+        Assertions.assertFalse(metrics.ignorePatterns.isEmpty());
+        Pattern p = metrics.ignorePatterns.get(0);
+        Assertions.assertEquals("/item/.*", p.pattern());
+        Assertions.assertTrue(p.matcher("/item/123").matches());
+    }
+
+    @Test
+    public void testHttpServerMetricsMatchPatterns() {
+        VertxConfig runtimeConfig = new VertxConfig();
+        runtimeConfig.matchPatterns = Optional.of(new ArrayList<>(Arrays.asList("/item/\\d+=/item/{id}")));
+        VertxHttpServerMetrics metrics = new VertxHttpServerMetrics(new SimpleMeterRegistry(), runtimeConfig);
+
+        Assertions.assertFalse(metrics.matchPatterns.isEmpty());
+        Map.Entry<Pattern, String> entry = metrics.matchPatterns.entrySet().iterator().next();
+        Assertions.assertEquals("/item/\\d+", entry.getKey().pattern());
+        Assertions.assertEquals("/item/{id}", entry.getValue());
+        Assertions.assertTrue(entry.getKey().matcher("/item/123").matches());
+    }
+}

--- a/extensions/micrometer/runtime/src/test/java/io/quarkus/micrometer/runtime/binder/vertx/VertxMetricsTagsTest.java
+++ b/extensions/micrometer/runtime/src/test/java/io/quarkus/micrometer/runtime/binder/vertx/VertxMetricsTagsTest.java
@@ -2,7 +2,9 @@ package io.quarkus.micrometer.runtime.binder.vertx;
 
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.regex.Pattern;
 
 import org.junit.jupiter.api.Assertions;
@@ -18,6 +20,7 @@ import io.vertx.core.http.HttpServerResponse;
 
 /**
  * Test tag creation
+ * Disabled on Java 8 because of Mocks
  */
 @DisabledOnJre(JRE.JAVA_8)
 public class VertxMetricsTagsTest {
@@ -26,7 +29,9 @@ public class VertxMetricsTagsTest {
     HttpServerResponse response;
 
     final List<Pattern> NO_IGNORE_PATTERNS = Collections.emptyList();
-    final List<Pattern> NO_MATCH_PATTERNS = Collections.emptyList();
+    final List<Pattern> ignorePatterns = Arrays.asList(Pattern.compile("/ignore.*"));
+
+    final Map<Pattern, String> NO_MATCH_PATTERNS = Collections.emptyMap();
 
     @BeforeEach
     public void initMocks() {
@@ -34,23 +39,81 @@ public class VertxMetricsTagsTest {
     }
 
     @Test
-    public void testParsePathNoIgnorePatterns() {
-        Assertions.assertEquals("/", VertxMetricsTags.parseUriPath(NO_MATCH_PATTERNS, NO_IGNORE_PATTERNS, "//"));
-        Assertions.assertEquals("/", VertxMetricsTags.parseUriPath(NO_MATCH_PATTERNS, NO_IGNORE_PATTERNS, ""));
-        Assertions.assertEquals("/path/with/no/leading/slash",
-                VertxMetricsTags.parseUriPath(NO_MATCH_PATTERNS, NO_IGNORE_PATTERNS, "path/with/no/leading/slash"));
-        Assertions.assertEquals("/path/with/query/string",
-                VertxMetricsTags.parseUriPath(NO_MATCH_PATTERNS, NO_IGNORE_PATTERNS, "/path/with/query/string?stuff"));
+    public void testParsePathDoubleSlash() {
+        RequestMetric requestMetric = new RequestMetric();
+        VertxMetricsTags.parseUriPath(requestMetric, NO_MATCH_PATTERNS, NO_IGNORE_PATTERNS, "//");
+        Assertions.assertEquals("/", requestMetric.path);
+        Assertions.assertTrue(requestMetric.measure);
+        Assertions.assertFalse(requestMetric.pathMatched);
     }
 
     @Test
-    public void testParsePathWithIgnorePatterns() {
-        List<Pattern> ignorePatterns = Arrays.asList(Pattern.compile("/ignore.*"));
+    public void testParseEmptyPath() {
+        RequestMetric requestMetric = new RequestMetric();
+        VertxMetricsTags.parseUriPath(requestMetric, NO_MATCH_PATTERNS, NO_IGNORE_PATTERNS, "");
+        Assertions.assertTrue(requestMetric.measure);
+        Assertions.assertFalse(requestMetric.pathMatched);
+        Assertions.assertEquals("/", requestMetric.path);
+    }
 
-        Assertions.assertNull(
-                VertxMetricsTags.parseUriPath(NO_MATCH_PATTERNS, ignorePatterns, "ignore/me/with/no/leading/slash"));
-        Assertions.assertNull(
-                VertxMetricsTags.parseUriPath(NO_MATCH_PATTERNS, ignorePatterns, "/ignore/me/with/query/string?stuff"));
+    @Test
+    public void testParsePathNoLeadingSlash() {
+        RequestMetric requestMetric = new RequestMetric();
+        VertxMetricsTags.parseUriPath(requestMetric, NO_MATCH_PATTERNS, NO_IGNORE_PATTERNS, "path/with/no/leading/slash");
+        Assertions.assertEquals("/path/with/no/leading/slash", requestMetric.path);
+        Assertions.assertTrue(requestMetric.measure);
+        Assertions.assertFalse(requestMetric.pathMatched);
+    }
+
+    @Test
+    public void testParsePathWithQueryString() {
+        RequestMetric requestMetric = new RequestMetric();
+        VertxMetricsTags.parseUriPath(requestMetric, NO_MATCH_PATTERNS, NO_IGNORE_PATTERNS, "/path/with/query/string?stuff");
+        Assertions.assertEquals("/path/with/query/string", requestMetric.path);
+        Assertions.assertTrue(requestMetric.measure);
+        Assertions.assertFalse(requestMetric.pathMatched);
+    }
+
+    @Test
+    public void testParsePathIgnoreNoLeadingSlash() {
+        RequestMetric requestMetric = new RequestMetric();
+        VertxMetricsTags.parseUriPath(requestMetric, NO_MATCH_PATTERNS, ignorePatterns, "ignore/me/with/no/leading/slash");
+        Assertions.assertEquals("/ignore/me/with/no/leading/slash", requestMetric.path);
+        Assertions.assertFalse(requestMetric.measure);
+        Assertions.assertFalse(requestMetric.pathMatched);
+    }
+
+    @Test
+    public void testParsePathIgnoreWithQueryString() {
+        RequestMetric requestMetric = new RequestMetric();
+        VertxMetricsTags.parseUriPath(requestMetric, NO_MATCH_PATTERNS, ignorePatterns, "/ignore/me/with/query/string?stuff");
+        Assertions.assertEquals("/ignore/me/with/query/string", requestMetric.path);
+        Assertions.assertFalse(requestMetric.measure);
+        Assertions.assertFalse(requestMetric.pathMatched);
+    }
+
+    @Test
+    public void testParsePathMatchReplaceNoLeadingSlash() {
+        final Map<Pattern, String> matchPatterns = new HashMap<>();
+        matchPatterns.put(Pattern.compile("/item/\\d+"), "/item/{id}");
+
+        RequestMetric requestMetric = new RequestMetric();
+        VertxMetricsTags.parseUriPath(requestMetric, matchPatterns, NO_IGNORE_PATTERNS, "item/123");
+        Assertions.assertEquals("/item/{id}", requestMetric.path);
+        Assertions.assertTrue(requestMetric.measure);
+        Assertions.assertTrue(requestMetric.pathMatched);
+    }
+
+    @Test
+    public void testParsePathMatchReplaceLeadingSlash() {
+        final Map<Pattern, String> matchPatterns = new HashMap<>();
+        matchPatterns.put(Pattern.compile("/item/\\d+"), "/item/{id}");
+
+        RequestMetric requestMetric = new RequestMetric();
+        VertxMetricsTags.parseUriPath(requestMetric, matchPatterns, NO_IGNORE_PATTERNS, "/item/123");
+        Assertions.assertEquals("/item/{id}", requestMetric.path);
+        Assertions.assertTrue(requestMetric.measure);
+        Assertions.assertTrue(requestMetric.pathMatched);
     }
 
     @Test

--- a/integration-tests/micrometer-jmx/pom.xml
+++ b/integration-tests/micrometer-jmx/pom.xml
@@ -26,14 +26,10 @@
             <artifactId>micrometer-registry-jmx</artifactId>
         </dependency>
 
-        <!-- JAX-RS -->
+        <!-- Do not use Resteasy! Deliberately focused on vertx-web alone -->
         <dependency>
             <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-resteasy</artifactId>
-        </dependency>
-        <dependency>
-            <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-resteasy-jackson</artifactId>
+            <artifactId>quarkus-vertx-web</artifactId>
         </dependency>
 
         <!-- Test Dependencies -->
@@ -69,20 +65,7 @@
       </dependency>
       <dependency>
         <groupId>io.quarkus</groupId>
-        <artifactId>quarkus-resteasy-deployment</artifactId>
-        <version>${project.version}</version>
-        <type>pom</type>
-        <scope>test</scope>
-        <exclusions>
-          <exclusion>
-            <groupId>*</groupId>
-            <artifactId>*</artifactId>
-          </exclusion>
-        </exclusions>
-      </dependency>
-      <dependency>
-        <groupId>io.quarkus</groupId>
-        <artifactId>quarkus-resteasy-jackson-deployment</artifactId>
+        <artifactId>quarkus-vertx-web-deployment</artifactId>
         <version>${project.version}</version>
         <type>pom</type>
         <scope>test</scope>

--- a/integration-tests/micrometer-jmx/src/main/java/io/quarkus/it/micrometer/jmx/MessageResource.java
+++ b/integration-tests/micrometer-jmx/src/main/java/io/quarkus/it/micrometer/jmx/MessageResource.java
@@ -4,14 +4,15 @@ import java.lang.management.ManagementFactory;
 import java.util.Set;
 
 import javax.management.*;
-import javax.ws.rs.GET;
-import javax.ws.rs.Path;
-import javax.ws.rs.PathParam;
 
 import io.micrometer.core.instrument.MeterRegistry;
 import io.micrometer.core.instrument.composite.CompositeMeterRegistry;
+import io.quarkus.vertx.web.Param;
+import io.quarkus.vertx.web.Route;
+import io.quarkus.vertx.web.RouteBase;
+import io.vertx.core.http.HttpMethod;
 
-@Path("/message")
+@RouteBase(path = "/message")
 public class MessageResource {
 
     private final MeterRegistry registry;
@@ -21,27 +22,24 @@ public class MessageResource {
         this.registry = registry;
     }
 
-    @GET
+    @Route(path = "ping", methods = HttpMethod.GET)
     public String message() {
         CompositeMeterRegistry compositeMeterRegistry = (CompositeMeterRegistry) registry;
         Set<MeterRegistry> subRegistries = compositeMeterRegistry.getRegistries();
         return subRegistries.iterator().next().getClass().getName();
     }
 
-    @GET
-    @Path("fail")
+    @Route(path = "fail", methods = HttpMethod.GET)
     public String fail() {
         throw new RuntimeException("Failed on purpose");
     }
 
-    @GET
-    @Path("item/{id}")
-    public String item(@PathParam("id") String id) {
+    @Route(path = "item/:id", methods = HttpMethod.GET)
+    public String item(@Param("id") String id) {
         return "return message with id " + id;
     }
 
-    @GET
-    @Path("mbeans")
+    @Route(path = "mbeans", methods = HttpMethod.GET)
     public String metrics() throws IntrospectionException, InstanceNotFoundException, ReflectionException {
         Set<ObjectName> mbeans = mBeanServer.queryNames(null, null);
         StringBuilder sb = new StringBuilder();

--- a/integration-tests/micrometer-jmx/src/test/java/io/quarkus/it/micrometer/jmx/JmxMetricsRegistryTest.java
+++ b/integration-tests/micrometer-jmx/src/test/java/io/quarkus/it/micrometer/jmx/JmxMetricsRegistryTest.java
@@ -24,7 +24,7 @@ class JmxMetricsRegistryTest {
     @Order(1)
     void testRegistryInjection() {
         given()
-                .when().get("/message")
+                .when().get("/message/ping")
                 .then()
                 .statusCode(200)
                 .body(containsString("io.micrometer.jmx.JmxMeterRegistry"));
@@ -49,7 +49,7 @@ class JmxMetricsRegistryTest {
     }
 
     @Test
-    @Order(3)
+    @Order(4)
     void testPathParameter() {
         given()
                 .when().get("/message/item/123")
@@ -64,6 +64,7 @@ class JmxMetricsRegistryTest {
         given()
                 .when().get("/message/mbeans")
                 .then()
+                .log().all()
                 .statusCode(200)
 
                 // JMX endpoint is returning a subset of mbean objects to inspect
@@ -78,7 +79,7 @@ class JmxMetricsRegistryTest {
                         "metrics:name=httpServerRequests.env.test.method.GET.outcome.SERVER_ERROR.registry.jmx.status.500.uri./message/fail"))
 
                 .body(containsString(
-                        "metrics:name=httpServerRequests.env.test.method.GET.outcome.SUCCESS.registry.jmx.status.200.uri./message"))
+                        "metrics:name=httpServerRequests.env.test.method.GET.outcome.SUCCESS.registry.jmx.status.200.uri./message/ping"))
                 .body(containsString(
                         "metrics:name=httpServerRequests.env.test.method.GET.outcome.SUCCESS.registry.jmx.status.200.uri./message/item/{id}"))
 

